data/src/main/java/org/apache/iceberg/data/orc/GenericOrcWriters.java

```diff
@@ -231,7 +234,13 @@ public void nonNullWrite(int rowId, String data, ColumnVector output) {
 
     @Override
     public void nonNullWrite(int rowId, ByteBuffer data, ColumnVector output) {
-      ((BytesColumnVector) output).setRef(rowId, data.array(), 0, data.array().length);
+      if (data != null && data.hasArray()) {
+        ((BytesColumnVector) output).setRef(rowId, data.array(), data.arrayOffset() + data.position(), data.remaining());
+      } else {
+        byte[] rawData = ByteBuffers.toByteArray(data);
+        int length = rawData == null ? 0 : rawData.length;
+        ((BytesColumnVector) output).setRef(rowId, rawData, 0, length);
```

Comment:
@rdblue please let me know if this is an appropriate use of `ByteBuffers.toByteArray` (since we know this is off-heap, I figure this brings it on heap in the way I was somewhat before).

And is the null check necessary? I saw that `toByteArray` returned null if `null` was passed in, and I didn't see any `@NonNull` kind of tags so I wasn't sure.

Reply:
The null check isn't needed because this is `nonNullWrite` so nulls have already been handled at this point.

Reply:
Haha didn't realize that. That's what I get for not looking at this PR for over a week 😄